File: modules/expo-headphone-detection/ios/ExpoHeadphoneDetectionModule.swift
import ExpoModulesCore
import AVFoundation

public class ExpoHeadphoneDetectionModule: Module {
  public func definition() -> ModuleDefinition {
    Name("ExpoHeadphoneDetection")

    AsyncFunction("getCurrentStatus") { (promise: Promise) in
      do {
        let status = getCurrentHeadphoneStatus()
        promise.resolve(status)
      } catch {
        promise.reject("HEADPHONE_DETECTION_ERROR", error.localizedDescription, error)
      }
    }

    AsyncFunction("startListening") { (promise: Promise) in
      do {
        // For now, just resolve - we can add real listening later
        promise.resolve(nil)
      } catch {
        promise.reject("START_LISTENING_ERROR", error.localizedDescription, error)
      }
    }

    AsyncFunction("stopListening") { (promise: Promise) in
      do {
        // For now, just resolve - we can add real listening later
        promise.resolve(nil)
      } catch {
        promise.reject("STOP_LISTENING_ERROR", error.localizedDescription, error)
      }
    }
  }

  private func getCurrentHeadphoneStatus() -> [String: Any] {
    let audioSession = AVAudioSession.sharedInstance()
    
    var isConnected = false
    var deviceType = "none"
    var deviceName = ""
    var confidence = 0.0

    // Check available audio routes
    let currentRoute = audioSession.currentRoute
    
    for output in currentRoute.outputs {
      switch output.portType {
      case .headphones, .headsetMic:
        isConnected = true
        deviceType = "wired"
        deviceName = output.portName
        confidence = 0.95
        break
      case .bluetoothA2DP, .bluetoothHFP, .bluetoothLE:
        isConnected = true
        deviceType = "bluetooth"
        deviceName = output.portName
        confidence = 0.90
        break
      default:
        continue
      }
    }

    return [
      "isConnected": isConnected,
      "deviceType": deviceType,
      "deviceName": deviceName,
      "confidence": confidence,
      "timestamp": Date().timeIntervalSince1970 * 1000,
      "metadata": [
        "platform": "ios",
        "routeDescription": currentRoute.description,
        "method": "AVAudioSession"
      ]
    ]
  }
}

File: modules/expo-headphone-detection/android/src/main/java/expo/modules/headphonedetection/ExpoHeadphoneDetectionModule.kt
package expo.modules.headphonedetection

import android.content.Context
import android.media.AudioManager
import android.media.AudioDeviceInfo
import android.os.Build
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.Promise

class ExpoHeadphoneDetectionModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("ExpoHeadphoneDetection")

    AsyncFunction("getCurrentStatus") { promise: Promise ->
      try {
        val status = getCurrentHeadphoneStatus()
        promise.resolve(status)
      } catch (error: Throwable) {
        promise.reject("HEADPHONE_DETECTION_ERROR", error.message, error)
      }
    }

    AsyncFunction("startListening") { promise: Promise ->
      try {
        // For now, just resolve - we can add real listening later
        promise.resolve(null)
      } catch (error: Throwable) {
        promise.reject("START_LISTENING_ERROR", error.message, error)
      }
    }

    AsyncFunction("stopListening") { promise: Promise ->
      try {
        // For now, just resolve - we can add real listening later
        promise.resolve(null)
      } catch (error: Throwable) {
        promise.reject("STOP_LISTENING_ERROR", error.message, error)
      }
    }
  }

  private val context
    get() = requireNotNull(appContext.reactContext)

  private fun getCurrentHeadphoneStatus(): Map<String, Any> {
    val audioManager = context.getSystemService(Context.AUDIO_SERVICE) as AudioManager
    
    var isConnected = false
    var deviceType = "none"
    var deviceName = ""
    var confidence = 0.0

    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
      val devices = audioManager.getDevices(AudioManager.GET_DEVICES_OUTPUTS)
      
      for (device in devices) {
        when (device.type) {
          AudioDeviceInfo.TYPE_WIRED_HEADPHONES,
          AudioDeviceInfo.TYPE_WIRED_HEADSET -> {
            isConnected = true
            deviceType = "wired"
            deviceName = device.productName?.toString() ?: "Wired Headphones"
            confidence = 0.95
            break
          }
          AudioDeviceInfo.TYPE_BLUETOOTH_A2DP,
          AudioDeviceInfo.TYPE_BLUETOOTH_SCO -> {
            isConnected = true
            deviceType = "bluetooth"
            deviceName = device.productName?.toString() ?: "Bluetooth Headphones"
            confidence = 0.90
            break
          }
        }
      }
    } else {
      // Fallback for older Android versions
      isConnected = audioManager.isWiredHeadsetOn || audioManager.isBluetoothA2dpOn
      if (isConnected) {
        deviceType = if (audioManager.isWiredHeadsetOn) "wired" else "bluetooth"
        deviceName = if (audioManager.isWiredHeadsetOn) "Wired Headphones" else "Bluetooth Headphones"
        confidence = 0.80
      }
    }

    return mapOf(
      "isConnected" to isConnected,
      "deviceType" to deviceType,
      "deviceName" to deviceName,
      "confidence" to confidence,
      "timestamp" to System.currentTimeMillis(),
      "metadata" to mapOf(
        "platform" to "android",
        "apiLevel" to Build.VERSION.SDK_INT,
        "method" to if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) "AudioDeviceInfo" else "AudioManager"
      )
    )
  }
}
